### PR TITLE
fix: use heuristics for build and deploy command as well

### PR DESCRIPTION
**Pull request by lukasholzer**
🎉 Thanks for submitting a pull request! 🎉

#### Summary

**Depends on: https://github.com/netlify/build/pull/5735**

Fixes https://linear.app/netlify/issue/CPLA-738/sites-created-in-the-cli-should-use-framework-detection

Run the framework detection before `ntl build` and `ntl deploy --build` as well to not need a `netlify.toml` in simple cases. (`ntl dev` already supported this)

---

For us to review and ship your PR efficiently, please perform the following steps:

- [ ] Open a [bug/issue](https://github.com/netlify/cli/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [ ] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and
      passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update or add documentation (if features were changed or added) 📝
- [ ] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**

### CODE REVIEW

File: src/utils/build-info.ts [hunk 1]
```diff
@@ -122,3 +123,34 @@ command = "${chosenSettings.devCommand}"
...
+    cachedConfig.config.build.commandOrigin = 'heuristics'
+  }
+
+  if (cachedConfig.config.build.publishOrigin === 'default') {
+    cachedConfig.config.build.publish = settings.dist
+    cachedConfig.config.build.publishOrigin = 'heuristics'
+  }
+
+  settings.plugins_recommended.forEach((plugin) => {
+    if (!cachedConfig.config.plugins.includes(plugin)) {
```
Comment (eduardoboucas) [inline src/utils/build-info.ts:150]: How does this check work? We're pushing objects such as `{ package: "package-name" }` to `cachedConfig.config.plugins`, so how can we use `includes` to check if a plugin is already there?

Comment (lukasholzer) [reply src/utils/build-info.ts:150]: that happens if you don't have types and everything is any 😬

**Review by eduardoboucas — APPROVED** [review]
It'd be nice to have some tests, but LGTM.

File: src/utils/build-info.ts [hunk 2]
```diff
@@ -122,3 +123,32 @@ command = "${chosenSettings.devCommand}"
     return chosenSettings
   }
 }
+
+/**
+ * Merges the settings from the heuristics with the cached config.
+ * Returns the updated cached config
+ * @param cachedConfig The cached config
```
Comment (eduardoboucas) [inline src/utils/build-info.ts:130]: I think this needs updating, since I don't see this param?

Comment (lukasholzer) [reply src/utils/build-info.ts:130]: oh good catch!

Comment (lukasholzer) [reply src/utils/build-info.ts]: fixed here: https://github.com/netlify/cli/pull/6729/commits/049f4a9c2819f2326e2289c593613defca3ea828

**Review by eduardoboucas — APPROVED** [review]
🚀

**Comment by lukasholzer** [comment]
@eduardoboucas I'm going to add an integration test to it as well :)

**Review by eduardoboucas — APPROVED** [review]

**Comment by ascorbic** [comment]
This is great! Should it be working for `serve` as well? I ask because I tried creating a new astro site and then ran `ntl serve` and it gave me a 404.

**Comment by flore2003** [comment]
Is there a way to overwrite or turn off this heuristic framework detection for the build and deploy commands? This new behavior breaks our current monorepo setup, because we have both Next.js and vite based React apps in it, and the heuristic detects next as the framework, even for the vite based React app. The `dev` command has an option to specify the framework, but neither build nor deploy seem to have it, if I see it correctly.

**Pull request merged**